man/tpm2_commit.1.md
% tpm2_commit(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_commit**(1) - Performs the first part of an ECC anonymous signing
operation.

# SYNOPSIS

**tpm2_commit** [*OPTIONS*]

# DESCRIPTION

**tpm2_commit**(1) - Performs the first part of an ECC anonymous signing
operation. The TPM will perform the point multiplications on the provided points
and return intermediate signing values. The signing key is an ECC key. The key
cannot be a sign+decrypt key and must have an anonymous signing scheme.
TPM_ALG_ECDAA is the only supported anonymous scheme.

# OPTIONS

  * **ARGUMENT**=_FILE_:

    Specify the input data used to derive the x coordinate of the basepoint.

  * **\--basepoint-y**=_FILE_:

    Specify the y coordinate of the basepoint.

  * **\--eccpoint-P**=_FILE_:

    Specify a point on the curve used by sign handle.

  * **\--eccpoint-K**=_FILE_:

    Output ECC point K ≔ \[ds\](x2, y2).

  * **\--eccpoint-L**=_FILE_:

    Output ECC point L ≔ \[r\](x2, y2).

  * **-u**, **\--public**=_FILE_:

    Output ECC point E ≔ [r]P1.

  * **-t**, **\--counter**=_FILE_

    Specify file path to save the least-significant 16 bits of commit count.

  * **-p**, **\--auth**=_AUTH_:

    The authorization value for the created object.

  * **-c**, **\--context**=_FILE_:

    Context object pointing to the the key used for signing. Either a file or a
    handle number. See section "Context Object Format".

  * **\--cphash**=_FILE_

    File path to record the hash of the command parameters. This is commonly
    termed as cpHash. NOTE: When this option is selected, The tool will not
    actually execute the command, it simply returns a cpHash.

## References

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

```bash
tpm2_createprimary -C o -c prim.ctx -Q

tpm2_create -C prim.ctx -c key.ctx -u key.pub -r key.priv -G ecc256:ecdaa

tpm2_commit -c key.ctx -t count.er \
--eccpoint-K K.bin --eccpoint-L L.bin -u E.bin
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_commit.c
/* SPDX-License-Identifier: BSD-3-Clause */
#include <string.h>

#include "files.h"
#include "log.h"
#include "object.h"
#include "tpm2.h"
#include "tpm2_tool.h"
#include "tpm2_auth_util.h"
#include "tpm2_options.h"

typedef struct tpm_ecephemeral_ctx tpm_ecephemeral_ctx;
struct tpm_ecephemeral_ctx {
    /*
     * Inputs
     */
    struct {
        const char *ctx_path;
        const char *auth_str;
        tpm2_loaded_object object;
    } signing_key;

    char *basepoint_x_coordinate_data_path;
    char *basepoint_y_data_path;
    char *eccpoint_M_data_path;

    TPM2B_ECC_POINT P1;
    TPM2B_SENSITIVE_DATA s2;
    TPM2B_ECC_PARAMETER y2;

    /*
     * Outputs
     */
    char *eccpoint_K_data_path;
    TPM2B_ECC_POINT *K;
    char *eccpoint_L_data_path;
    TPM2B_ECC_POINT *L;
    char *eccpoint_E_data_path;
    TPM2B_ECC_POINT *E;
    char *commit_counter_path;
    uint16_t counter;

    /*
     * Parameter hashes
     */
    char *cp_hash_path;
    TPM2B_DIGEST *cphash;
    TPM2B_DIGEST cp_hash;
    bool is_command_dispatch;
};

static tpm_ecephemeral_ctx ctx;

static tool_rc commit(ESYS_CONTEXT *ectx) {

    return tpm2_commit(ectx, &ctx.signing_key.object, &ctx.P1, &ctx.s2, &ctx.y2,
        &ctx.K, &ctx.L, &ctx.E, &ctx.counter, ctx.cphash);
}

static tool_rc process_outputs(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Outputs that do not require TPM2_CC_<command> dispatch
     */
    bool is_file_op_success = true;
    if (ctx.cp_hash_path) {
        is_file_op_success = files_save_digest(&ctx.cp_hash, ctx.cp_hash_path);

        if (!is_file_op_success) {
            return tool_rc_general_error;
        }
    }

    tool_rc rc = tool_rc_success;
    if (!ctx.is_command_dispatch) {
        return rc;
    }

    /*
     * 2. Outputs generated after TPM2_CC_<command> dispatch
     */
    is_file_op_success = files_save_ecc_point(ctx.K, ctx.eccpoint_K_data_path);
    if (!is_file_op_success) {
        LOG_ERR("Failed to write out the ECC point K");
        return tool_rc_general_error;
    }

    is_file_op_success = files_save_ecc_point(ctx.L, ctx.eccpoint_L_data_path);
    if (!is_file_op_success) {
        LOG_ERR("Failed to write out the ECC point L");
        return tool_rc_general_error;
    }

    is_file_op_success = files_save_ecc_point(ctx.E, ctx.eccpoint_E_data_path);
    if (!is_file_op_success) {
        LOG_ERR("Failed to write out the ECC point E");
        return tool_rc_general_error;
    }

    FILE *fp = fopen(ctx.commit_counter_path, "wb");
    is_file_op_success = files_write_16(fp, ctx.counter);
    fclose(fp);
    if (!is_file_op_success) {
        LOG_ERR("Failed to write out the ECC commit count");
        return tool_rc_general_error;
    }

    return tool_rc_success;
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    /*
     * 1. Object and auth initializations
     */

    /*
     * 1.a Add the new-auth values to be set for the object.
     */

    /*
     * 1.b Add object names and their auth sessions
     */
    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.signing_key.ctx_path,
            ctx.signing_key.auth_str, &ctx.signing_key.object, false,
            TPM2_HANDLES_FLAGS_TRANSIENT|TPM2_HANDLES_FLAGS_PERSISTENT);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Restore auxiliary sessions
     */

    /*
     * 3. Command specific initializations
     */

    bool result = true;
    if (ctx.basepoint_x_coordinate_data_path) {
        result = files_load_ecc_point(ctx.eccpoint_M_data_path, &ctx.P1);
    }
    if (!result) {
        LOG_ERR("Failed to load input ECC point P1");
        return tool_rc_general_error;
    }

    if (ctx.basepoint_y_data_path) {
        result = files_load_ecc_parameter(ctx.basepoint_y_data_path, &ctx.y2);
    }
    if (!result) {
        LOG_ERR("Failed to load input ECC parameter y2");
        return tool_rc_general_error;
    }

    /*
     * 4. Configuration for calculating the pHash
     */
    ctx.cphash = ctx.cp_hash_path ? &ctx.cp_hash : 0;

    /*
     * 4.a Determine pHash length and alg
     */

    /*
     * 4.b Determine if TPM2_CC_<command> is to be dispatched
     */
    ctx.is_command_dispatch = ctx.cp_hash_path ? false : true;

    return rc;
}

static tool_rc check_options(void) {

    if (!ctx.signing_key.ctx_path) {
        LOG_ERR("Specify a signing key");
        return tool_rc_option_error;
    }

    if (ctx.basepoint_y_data_path && !ctx.basepoint_x_coordinate_data_path) {
        LOG_ERR("Specify parameter data for basepoint X coordinate derivation");
        return tool_rc_option_error;
    }

    return tool_rc_success;
}

static bool on_args(int argc, char **argv) {

    if (argc > 1) {
        LOG_ERR("Specify single argument with file input containing the "
                "octet array used to derive x-coordinate of a base point");
        return false;
    }

    ctx.basepoint_x_coordinate_data_path = strcmp("-", argv[0]) ? argv[0] : 0;

    return files_load_bytes_from_buffer_or_file_or_stdin(0,
        ctx.basepoint_x_coordinate_data_path, &ctx.s2.size, ctx.s2.buffer);
}

static bool on_option(char key, char *value) {

    switch (key) {
        case 'p':
            ctx.signing_key.auth_str = value;
            break;
        case 'c':
            ctx.signing_key.ctx_path = value;
            break;
        case 0:
            ctx.basepoint_y_data_path = value;
            break;
        case 1:
            ctx.eccpoint_M_data_path = value;
            break;
        case 2:
            ctx.eccpoint_K_data_path = value;
            break;
        case 3:
            ctx.eccpoint_L_data_path = value;
            break;
        case 'u':
            ctx.eccpoint_E_data_path = value;
            break;
        case 't':
            ctx.commit_counter_path = value;
            break;
        case 4:
            ctx.cp_hash_path = value;
            break;
    };

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static struct option topts[] = {
      { "auth",        required_argument, 0, 'p' },
      { "context",     required_argument, 0, 'c' },
      { "basepoint-y", required_argument, 0,  0  },
      { "eccpoint-P",  required_argument, 0,  1  },
      { "eccpoint-K",  required_argument, 0,  2  },
      { "eccpoint-L",  required_argument, 0,  3  },
      { "public",      required_argument, 0, 'u' },
      { "counter",     required_argument, 0, 't' },
      { "cphash",      required_argument, 0,  4  },
    };

    *opts = tpm2_options_new("p:c:t:u:", ARRAY_LEN(topts), topts,
            on_option, on_args, 0);

    return *opts != 0;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    /*
     * 1. Process options
     */
    tool_rc rc = check_options();
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Process inputs
     */
    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 3. TPM2_CC_<command> call
     */
    rc = commit(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 4. Process outputs
     */
    return process_outputs(ectx);
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Free objects
     */

    /*
     * 2. Close authorization sessions
     */

    /*
     * 3. Close auxiliary sessions
     */

    return tool_rc_success;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("commit", tpm2_tool_onstart, tpm2_tool_onrun,
    tpm2_tool_onstop, 0)
